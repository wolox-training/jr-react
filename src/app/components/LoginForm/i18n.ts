import i18next from 'i18next';

i18next.addResources('es', 'Register', {
  firstName: 'Nombre',
  lastName: 'Apellido',
  logoAlt: 'logo Wolox',
  messageSuccess: 'Usuario registrado',
  passwordConfirmation: 'Confirmación de Contraseña'
});

i18next.addResources('en', 'Register', {
  firstName: 'Name',
  lastName: 'Surname',
  logoAlt: 'logo Wolox',
  messageSuccess: 'Registered user',
  passwordConfirmation: 'Repeat Password'
});
