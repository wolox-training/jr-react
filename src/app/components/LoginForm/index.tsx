/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';

import Input from '~components/Input';
import { emailRegex } from '~utils/inputValidations';
import { AUTH_FIELDS } from '~constants/fields';
import { ContentForm } from '~utils/types';
import { LoginBody } from '~screens/Login/types';

import styles from './styles.module.scss';

function LoginForm({ onSubmit, isLoading }: ContentForm) {
  const { register, handleSubmit, errors, formState } = useForm<LoginBody>({
    mode: 'all'
  });

  const submit = handleSubmit(data => {
    onSubmit(data);
  });

  return (
    <>
      <form className={styles.body} onSubmit={submit}>
        <Input
          labelText={i18next.t('FormAuth:email')}
          name={AUTH_FIELDS.email}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' },
            pattern: {
              value: emailRegex,
              message: 'FormValidations:errorEmail'
            }
          })}
          type="email"
          errorMessage={errors.email?.message}
        />
        <Input
          labelText={i18next.t('FormAuth:password')}
          name={AUTH_FIELDS.password}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' }
          })}
          type="password"
          errorMessage={errors.password?.message}
        />
        <button type="submit" disabled={!formState.isValid || isLoading} className="button btn-green">
          {i18next.t('FormAuth:btnLogin')}
        </button>
      </form>
      <div className={styles.footer}>
        <button type="button" className="button btn-white">
          {i18next.t('FormAuth:btnRegister')}
        </button>
      </div>
    </>
  );
}

export default LoginForm;
