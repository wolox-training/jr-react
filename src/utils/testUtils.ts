import { screen, fireEvent } from '@testing-library/react';

export const setInputByLabelText = (name: string, value: string) => {
  fireEvent.input(screen.getByLabelText(name), {
    target: {
      value
    }
  });
};
