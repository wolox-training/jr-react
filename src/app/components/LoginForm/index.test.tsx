import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import { setInputByLabelText } from '~utils/testUtils';

import LoginForm from './index';

const I18N_KEYS = {
  email: 'FormAuth:email',
  password: 'FormAuth:password',
  buttonLogin: 'FormAuth:btnLogin'
};

const USER_TEST = {
  email: 'federico.gomez@test.com',
  password: 'asdf1234'
};

const mockOnSubmit = jest.fn();
describe('Login Form Component', () => {
  it('Renders successfully', () => {
    const { asFragment } = render(<LoginForm onSubmit={mockOnSubmit} isLoading={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('If the form is not completed correctly, it is not sent', async () => {
    render(<LoginForm onSubmit={mockOnSubmit} isLoading={false} />);

    const buttonLogin = screen.getByRole('button', { name: I18N_KEYS.buttonLogin });

    await waitFor(() => {
      fireEvent.click(buttonLogin);
    });

    const alertErrorShow = 2;
    expect(await screen.findAllByRole('alert')).toHaveLength(alertErrorShow);
    expect(mockOnSubmit).not.toBeCalled();
  });

  test('Successful login', async () => {
    render(<LoginForm onSubmit={mockOnSubmit} isLoading={false} />);
    await waitFor(() => {
      setInputByLabelText(I18N_KEYS.email, USER_TEST.email);
      setInputByLabelText(I18N_KEYS.password, USER_TEST.password);
    });

    const buttonLogin = screen.getByRole('button', { name: I18N_KEYS.buttonLogin });

    await waitFor(() => {
      fireEvent.click(buttonLogin);
    });

    expect(mockOnSubmit).toHaveBeenCalled();
  });
});
