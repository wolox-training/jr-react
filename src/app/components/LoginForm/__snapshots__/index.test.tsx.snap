// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Component Renders successfully 1`] = `
<DocumentFragment>
  <form
    class="body"
    name="loginForm"
  >
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="email"
      >
        FormAuth:email
      </label>
      <input
        class="full-width input"
        id="email"
        name="email"
        type="email"
      />
    </div>
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="password"
      >
        FormAuth:password
      </label>
      <input
        class="full-width input"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <button
      class="button btn-green"
      type="submit"
    >
      FormAuth:btnLogin
    </button>
  </form>
  <div
    class="footer"
  >
    <button
      class="button btn-white"
      type="button"
    >
      FormAuth:btnRegister
    </button>
  </div>
</DocumentFragment>
`;
